### PR TITLE
Fix multiple strands on call stack issue

### PR DESCRIPTION
**Pull request by praveennadarajah**
## Purpose
This PR will fix multiple strands on call stack issue.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/27913

### CODE REVIEW

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java [hunk 1]
```diff
@@ -152,17 +153,35 @@ void setBreakpointsList(String path, Breakpoint[] breakpointsList) {
...
+                    breakPointThreads.put(threadReference.uniqueID(), threadReference);
+                }
+            } catch (AbsentInformationException | IncompatibleThreadStateException e) {
+                LOGGER.error(e.getMessage());
             }
         }
         return breakPointThreads;
     }
 
+    boolean isBalStrand(ThreadReference threadReference)
```
Comment (NipunaRanasinghe) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java:172]: Shall we make this private and static?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java [hunk 2]
```diff
@@ -152,17 +153,35 @@ void setBreakpointsList(String path, Breakpoint[] breakpointsList) {
...
+            throws IncompatibleThreadStateException, AbsentInformationException {
+        List<StackFrame> stackFrames = threadReference.frames();
+
+        if (stackFrames.isEmpty()) {
+            return false;
+        }
+        if (stackFrames.get(0).location() == null) {
+            return false;
+        }
```
Comment (NipunaRanasinghe) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java:181]: Can we merge these if conditions?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java [hunk 3]
```diff
@@ -152,17 +153,35 @@ void setBreakpointsList(String path, Breakpoint[] breakpointsList) {
...
+        return stackFrames.get(0).location().sourceName().endsWith(".bal");
```
Comment (NipunaRanasinghe) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java:182]: Can `sourceName` be null in here?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java (hunk 1 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java]: Made ti private static.

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java (hunk 2 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java]: Done

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java (hunk 3 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java]: Here I return false when an exception occurred when evaluating the expression.

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java [hunk 4]
```diff
@@ -153,16 +154,25 @@ void setBreakpointsList(String path, Breakpoint[] breakpointsList) {
...
                 breakPointThreads.put(threadReference.uniqueID(), threadReference);
             }
         }
         return breakPointThreads;
     }
 
+    private static boolean isBalStrand(ThreadReference threadReference) {
+        try {
+            return threadReference.frames().get(0).location().sourceName().endsWith(BAL_FILE_EXT);
+        } catch (AbsentInformationException | IncompatibleThreadStateException e) {
```
Comment (NipunaRanasinghe) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java:171]:
```suggestion
        } catch (Exception e) {
```

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java (hunk 3 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java]: https://github.com/ballerina-platform/ballerina-lang/pull/28238#discussion_r565030890

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java (hunk 4 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java]: Done

**Review by NipunaRanasinghe — APPROVED** [review]

**Review by NipunaRanasinghe — COMMENTED** [review]
Also, IMO better to add test case(s) to to ensure that we always get the active strands. WDYT?

**Comment by praveennadarajah** [comment]
+1 for adding tes case(s). I have created an issue to track this https://github.com/ballerina-platform/ballerina-lang/issues/28245

**Pull request merged**